### ISSUE TITLE
Subcommand to rebuild documentation when sources change

### ISSUE THREAD

**Issue by Bzero** [lawmurray/doxide#59]
When writing documentation I find it very helpful to get immediate feedback on how it ends up looking, especially if one does not yet have lots of experience with a particular tool. Some tools have subcommands for this purpose to automatically rebuild when changes in its source occur (e.g. `mkdocs serve`, `mdbook watch` or `typst watch`).

I would suggest to add a similar subcommand to doxide (e.g. `doxide watch`) which would watch the documentation's source files and rebuild it on changes. 

If this is considered a useful feature I would be glad to try to implement it.

**Comment by lawmurray** [comment]
Thanks @Bzero, great idea I think! I would find this very useful as well, I usually have `mkdocs serve` running but have to manually run `doxide build` on each change, so a `doxide watch` would certainly improve my workflow, giving immediate feedback as you say.

Very happy for you to have a go at implementing this. Let me know if you need any pointers to get started.

**Comment by lawmurray** [comment]
Sounds good @Bzero. One simple approach that comes to mind is to just keep track of last modified times and periodically (every few seconds?) scan for new and modified files. That would avoid additional dependencies and be cross-platform. An added layer would be to keep a hash of the file contents, and if the last modified time is updated, use the hash to check whether the file has actually changed (`rclone` works this way).

If changes are found it might currently be necessary to trigger the whole build again, as the data structures are not set up for incremental changes.

So a minimum viable approach here might just be periodic scan and full rebuild on changes. But interested to see what you come up with!

**Comment by Bzero** [comment]
While I would still think using the os provided method (like inotify on linux) would be the more scalable and more responsive way, I agree that with you that periodically checking file modification time is simpler to start with so I decided to take that way for now.

**Comment by lawmurray** [comment]
Thanks for the PR @Bzero! I'll get it reviewed, but had a quick look already and looking good.

I agree with you that something like inotify would be more scalable and responsive. My thinking is the scan approach gives a baseline that's portable across platforms, and a followup iteration could be using a portable library (if you found one) for an event driven approach, or if not can add system-specific approaches like inotify one at a time.

So, for example, if you see a clear way forward to add inotify for Linux but you're blocked on how to support other platforms, it would be acceptable now to add a Linux specific improvement, knowing that other platforms are still supported with the baseline approach.

**Comment by lawmurray** [comment]
Thanks again for the contribution @Bzero, will close now that #66 is merged.

**Comment by lawmurray** [comment]
@Bzero Related to this, I stumbled upon [Sane C++ Libraries](https://pagghiu.github.io/SaneCppLibraries/index.html), which seems to provide [portable file system notifications](https://pagghiu.github.io/SaneCppLibraries/library_file_system_watcher.html) that could be used to improve the new watch command.

**Comment by Bzero** [comment]
Thanks for the hint @lawmurray, I will look into it.

**Comment by lawmurray** [comment]
Another candidate: [libfswatch](https://emcrisostomo.github.io/fswatch). I'll open a new ticket for this task actually, for visibility, given that this one is closed.
